src/circuit_writer/ir.rs

```diff
@@ -0,0 +1,1119 @@
+                            _ => Err(Error::new(
+                                "compile-stmt",
+                                ErrorKind::UnexpectedError("expected array"),
+                                stmt.span,
+                            ))?,
+                        };
+
+                        // compute the size of each element in the array
+                        let len = self.size_of(&elem_type);
+
+                        for idx in 0..array_len {
+                            // compute the real index
+                            let idx = idx as usize;
+                            let start = idx * len;
+
+                            fn_env.nest();
+
+                            // add the variable to the inner enviroment corresponding
+                            // to iterator[idx]
+                            let indexed_var = iterator_var.narrow(start, len).value(self, fn_env);
+                            let var_info =
+                                VarInfo::new(indexed_var.clone(), false, Some(*elem_type.clone()));
+                            self.add_local_var(fn_env, var.value.clone(), var_info)?;
+
+                            self.compile_block(fn_env, body)?;
+
+                            fn_env.pop();
+                        }
+                    }
+                }
+            }
+            StmtKind::Expr(expr) => {
+                // compute the expression
+                let var = self.compute_expr(fn_env, expr)?;
+
+                // make sure it does not return any value.
+                assert!(var.is_none());
+            }
+            StmtKind::Return(expr) => {
+                let var = self
+                    .compute_expr(fn_env, expr)?
+                    .ok_or_else(|| self.error(ErrorKind::CannotComputeExpression, stmt.span))?;
+
+                // we already checked in type checking that this is not an early return
+                return Ok(Some(var));
+            }
+            StmtKind::Comment(_) => (),
+        }
+
+        Ok(None)
+    }
+
+    /// Same as circuit_writer::compile_block
+    /// might return something?
+    fn compile_block(&mut self, fn_env: &mut FnEnv, stmts: &[Stmt]) -> Result<Option<Var>> {
+        fn_env.nest();
+        for stmt in stmts {
+            let res = self.compile_stmt(fn_env, stmt)?;
+            if let Some(var) = res {
+                // a block doesn't return a pointer, only values
+                let var = var.value(self, fn_env);
+
+                // we already checked for early returns in type checking
+                return Ok(Some(var));
+            }
+        }
+        fn_env.pop();
+        Ok(None)
+    }
+
+    pub fn compile_hint_function_call(
+        &mut self,
+        function: &FunctionDef,
+        args: Vec<crate::circuit_writer::fn_env::VarInfo<B::Field, B::Var>>,
+    ) -> Result<Vec<crate::var::Value<B>>> {
+        assert!(!function.is_main());
```

Comment:
Do we actually check that we are calling a hint function and not a non hint one?

Reply:
Also, a hint should only be able to call hint function, right?

Reply:
yes, only hint functions are allowed.